### ISSUE TITLE
Spack install from custom repo defined in an environment on Darwin broken

### ISSUE THREAD

**Issue by pbrady** [LLNL/spack#45285]
### Steps to reproduce

A git bisect identified commit 95cf341b5064608650ccdd67f5fed35ba6c56a7d associated with PR #45053 as the culprit for breaking my workflow.  I'm not sure how reproducible this is but here is my setup:

I have an evironment with the following `spack.yaml`
```yaml
spack:
  # add package specs to the `specs` list
  specs:
  - riot ^singularity-eos@git.91ee6924a607899a65700aacc94e7a580b2bd5e1=1.8.5 ~eospac ^openmpi ^ports-of-call@git.v1.6.0
  view: true
  concretizer:
    unify: true
  repos:
  - /Users/ptb/gitrepos/riot/spack-repo
```

The contents of `/Users/ptb/gitrepos/riot/spack-repo` are:
```console
> tree spack-repo
spack-repo
├── packages
│   ├── parthenon
│   │   └── package.py
│   └── riot
│       └── package.py
└── repo.yaml
```

My parthenon `package.py` is very simple:
```python
from spack import *
from spack.pkg.builtin.parthenon import Parthenon


class Parthenon(Parthenon):
    """Additional tagged versions."""

    version("24.03.2", commit="bd615b7fdf806b14fbcddfa2dfc3c4347e72bad2")
    version("24.03.1", commit="dedc15f634b3b9bd60a919f7781ac8d376b2ca98")
```
Prior to #45053 a `spack concretize -fU` followed by `spack install --only dependencies` completes without error.  After #45053, I get the following error:

```console
==> Installing parthenon-24.03.2-3iwkmmtbprhumsnhg6vb2w5lhax47j2a [43/43]
==> No binary for parthenon-24.03.2-3iwkmmtbprhumsnhg6vb2w5lhax47j2a found: installing from source
==> Error: ModuleNotFoundError: No module named 'spack.pkg.riot'
==> Error: Failed to install parthenon due to TypeError: expected str, bytes or os.PathLike object, not NoneType
```

### Error message

```console
lib/spack/spack/installer.py:1993 ==> [2024-07-17-18:34:04.569382] Processing parthenon-24.03.2-3iwkmmtbprhumsnhg6vb2w5lhax47j2a: task=priority=0, status=dequeued, start=1721259244.49623, #dependencies=5
lib/spack/spack/installer.py:1481 ==> [2024-07-17-18:34:04.569626] Acquiring a write lock on parthenon-24.03.2-3iwkmmtbprhumsnhg6vb2w5lhax47j2a with timeout 1.000ns
lib/spack/spack/installer.py:1510 ==> [2024-07-17-18:34:04.569721] parthenon-24.03.2-3iwkmmtbprhumsnhg6vb2w5lhax47j2a is now write locked
lib/spack/spack/installer.py:1637 ==> [2024-07-17-18:34:04.654679] Installing parthenon-24.03.2-3iwkmmtbprhumsnhg6vb2w5lhax47j2a [43/43]
lib/spack/spack/config.py:1123 ==> [2024-07-17-18:34:04.654937] Reading config from file /Users/ptb/gitrepos/spack/etc/spack/defaults/mirrors.yaml
lib/spack/spack/installer.py:1651 ==> [2024-07-17-18:34:04.655946] No binary for parthenon-24.03.2-3iwkmmtbprhumsnhg6vb2w5lhax47j2a found: installing from source
lib/spack/spack/installer.py:1813 ==> [2024-07-17-18:34:04.656080] Creating the installation directory /Users/ptb/gitrepos/spack/opt/spack/darwin-ventura-m1/apple-clang-15.0.0/parthenon-24.03.2-3iwkmmtbprhumsnhg6vb2w5lhax47j2a
lib/spack/spack/util/executable.py:233 ==> [2024-07-17-18:34:04.660343] '/usr/bin/git' '-C' '/Users/ptb/gitrepos/spack' 'rev-parse' 'HEAD'
lib/spack/spack/error.py:52 ==> [2024-07-17-18:34:04.939598] Error: ModuleNotFoundError: No module named 'spack.pkg.riot'
lib/spack/spack/installer.py:1850 ==> [2024-07-17-18:34:04.939795] Flagging parthenon-24.03.2-3iwkmmtbprhumsnhg6vb2w5lhax47j2a as failed: expected str, bytes or os.PathLike object, not NoneType
lib/spack/spack/installer.py:1866 ==> [2024-07-17-18:34:04.942790] No build task for riot-develop-2rlqrvv3avkseky4554fehxqigxh7gve to skip since parthenon-24.03.2-3iwkmmtbprhumsnhg6vb2w5lhax47j2a failed
lib/spack/spack/installer.py:2161 ==> [2024-07-17-18:34:04.942879] Error: Failed to install parthenon due to TypeError: expected str, bytes or os.PathLike object, not NoneType
lib/spack/spack/installer.py:1496 ==> [2024-07-17-18:34:04.945610] Downgrading to a read lock on parthenon-24.03.2-3iwkmmtbprhumsnhg6vb2w5lhax47j2a with timeout 1.000ns
```

### Information on your system

* **Spack:** 0.23.0.dev0 (95cf341b5064608650ccdd67f5fed35ba6c56a7d)
* **Python:** 3.12.4
* **Platform:** darwin-ventura-m1
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by pbrady** [comment]
This actually looks to be a duplicate of #45197.  I may have forgotten to pull the latest from `develop` before starting my `bisect`.  Will double check and close if duplicate.

**Comment by alalazo** [comment]
Duplicate of #45197

**Comment by alalazo** [comment]
@pbrady It looks like you miss the latest commits:

- #45233 (which makes the error message more readable)
- #45229 (which fixes this issue)
